src/backend/distributed/metadata/node_metadata.c

```diff
@@ -2157,6 +2111,79 @@ AddNodeMetadata(char *nodeName, int32 nodePort,
 }
 
 
+/*
+ * AddNodeMetadataViaMetadataContext does the same thing as AddNodeMetadata but
+ * make use of metadata sync context to send commands to workers to support both
+ * transactional and nontransactional sync modes.
+ */
+static int
+AddNodeMetadataViaMetadataContext(char *nodeName, int32 nodePort,
+								  NodeMetadata *nodeMetadata, bool *nodeAlreadyExists)
+{
+	bool localOnly = true;
+	int nodeId = AddNodeMetadata(nodeName, nodePort, nodeMetadata, nodeAlreadyExists,
+								 localOnly);
+
+	/* do nothing as the node already exists */
+	if (*nodeAlreadyExists)
+	{
+		return nodeId;
+	}
+
+	/*
+	 * Create metadata sync context that is used throughout node addition
+	 * and activation if necessary.
+	 */
+	WorkerNode *node = FindWorkerNodeAnyCluster(nodeName, nodePort);
+	List *nodeList = list_make1(node);
+	bool collectCommands = false;
+	bool nodesAddedInSameTransaction = true;
+	MetadataSyncContext *context = CreateMetadataSyncContext(nodeList, collectCommands,
+															 nodesAddedInSameTransaction);
+
+	if (EnableMetadataSync)
+	{
+		/* send the delete command to all primary nodes with metadata */
+		char *nodeDeleteCommand = NodeDeleteCommand(node->nodeId);
+		SendOrCollectCommandListToMetadataNodes(context, list_make1(nodeDeleteCommand));
+
+		/* finally prepare the insert command and send it to all primary nodes */
+		uint32 primariesWithMetadata = CountPrimariesWithMetadata();
+		if (primariesWithMetadata != 0)
+		{
+			char *nodeInsertCommand = NodeListInsertCommand(nodeList);
+			SendOrCollectCommandListToMetadataNodes(context,
+													list_make1(nodeInsertCommand));
+		}
+	}
+
+	/*
+	 * After adding new node, if the node did not already exist, we will activate
+	 * the node.
+	 * If the worker is not marked as a coordinator, check that
+	 * the node is not trying to add itself
+	 */
+	if (node != NULL &&
+		node->groupId != COORDINATOR_GROUP_ID &&
+		node->nodeRole != SecondaryNodeRoleId() &&
+		IsWorkerTheCurrentNode(node))
+	{
+		ereport(ERROR, (errmsg("Node cannot add itself as a worker."),
+						errhint(
+							"Add the node as a coordinator by using: "
+							"SELECT citus_set_coordinator_host('%s', %d);",
+							node->workerName, node->workerPort)));
+	}
+
+	ActivateNodeList(context);
```

Comment:
why do we activate the nodes? This function is NOT supposed to activate nodes. Otherwise, we have this complexity regarding secondary nodes.

Reply:
`citus_add_node` also activates nodes if they do not exist before. I think we should address that in another PR.

Reply:
Wow, that I didn't know. We seem to heavily rely on `NodeIsPrimary `. I think it'd be good to see such refactor after this PR.